pytorch_lightning/models/trainer.py

```diff
@@ -173,13 +174,13 @@ def __init__(self,
 
             # set the correct cuda visible devices (using pci order)
             os.environ["CUDA_DEVICE_ORDER"] = "PCI_BUS_ID"
-            os.environ["CUDA_VISIBLE_DEVICES"] = ','.join([str(x) for x in self.data_parallel_device_ids])
-            print(f'VISIBLE GPUS: {os.environ["CUDA_VISIBLE_DEVICES"]}')
+            os.environ["CUDA_VISIBLE_DEVICES"] = ','.join([str(x) for x in
+                                                           self.data_parallel_device_ids])
+            print('VISIBLE GPUS: %r' % os.environ["CUDA_VISIBLE_DEVICES"])
 
         # make DP and DDP mutually exclusive
         # single GPU will also use DP with devices=[0]
-        have_gpus = self.data_parallel_device_ids is not None and len(self.data_parallel_device_ids) > 0
-        if have_gpus:
+        if self.data_parallel_device_ids:
```

Comment:
unfortunately the check needs to be:

```
self.data_parallel_device_ids is not None and len(self.data_parallel_device_ids) > 0
```

Case 1: is not None (user didn't pass gpus), so skip statement
Case 2: user did pass gpus AND it's more than a single GPU then we do whatever backend the user wants.
Case 3: use passed in a single GPU. In this case, we don't want to do DDP. We want to keep it as DP because DDP won't work well with a single GPU.

Reply:
I feel lost, what is the case when you want to enter the if block? not empty array?
```
Python 3.6.8 (default, Jan 14 2019, 11:02:34) 
var = None
True if var else False
Out[3]: False
var = []
True if var else False
Out[5]: False
var = [2]
True if var else False
Out[7]: True
```

Reply:
> 1 GPUs.  But if you don't do the None check before doing the check, it'll crash...

You want:
```python
if more_than_1_gpus:
    # enter
```

But can't do this:
```python
more_than_1_gpus = len(self.data_parallel_device_ids) > 0
```

because self.data_parallel_device_ids is None, so it'll crash.

Thus you have to check for that first as well (which is the case when Non GPU ids are passed)